standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/DatabaseProduct.java

```diff
@@ -20,71 +20,666 @@
+
+  private static enum DbType {DERBY, MYSQL, POSTGRES, ORACLE, SQLSERVER, CUSTOM, UNDEFINED};
+  public DbType dbType;
+  
+  // Singleton instance
+  private static DatabaseProduct theDatabaseProduct;
+
+  Configuration myConf;
+  /**
+   * Protected constructor for singleton class
+   * @param id
+   */
+  protected DatabaseProduct() {}
+
+  public static final String DERBY_NAME = "derby";
+  public static final String SQL_SERVER_NAME = "microsoft sql server";
+  public static final String MYSQL_NAME = "mysql";
+  public static final String POSTGRESQL_NAME = "postgresql";
+  public static final String ORACLE_NAME = "oracle";
+  public static final String UNDEFINED_NAME = "other";
+  
   /**
    * Determine the database product type
    * @param productName string to defer database connection
    * @return database product type
    */
-  public static DatabaseProduct determineDatabaseProduct(String productName) throws SQLException {
-    if (productName == null) {
-      return OTHER;
+  public static DatabaseProduct determineDatabaseProduct(String productName, Configuration c) {
+    DbType dbt;
+
+    if (theDatabaseProduct != null) {
+      Preconditions.checkState(theDatabaseProduct.dbType == getDbType(productName));
+      return theDatabaseProduct;
     }
+
+    // This method may be invoked by concurrent connections
+    synchronized (DatabaseProduct.class) {
+
+      if (productName == null) {
+        productName = UNDEFINED_NAME;
+      }
+
+      dbt = getDbType(productName);
+
+      // Check for null again in case of race condition
+      if (theDatabaseProduct == null) {
+        final Configuration conf = c!= null ? c : MetastoreConf.newMetastoreConf();
+        // Check if we are using an external database product
+        boolean isExternal = MetastoreConf.getBoolVar(conf, ConfVars.USE_CUSTOM_RDBMS);
+
+        if (isExternal) {
+          // The DatabaseProduct will be created by instantiating an external class via
+          // reflection. The external class can override any method in the current class
+          String className = MetastoreConf.getVar(conf, ConfVars.CUSTOM_RDBMS_CLASSNAME);
+          
+          if (className != null) {
+            try {
+              theDatabaseProduct = (DatabaseProduct)
+                  ReflectionUtils.newInstance(Class.forName(className), conf);
+  
+              LOG.info(String.format("Using custom RDBMS %s. Overriding DbType: %s", className, dbt));
+              dbt = DbType.CUSTOM;
+            }catch (Exception e) {
+              LOG.warn("Caught exception instantiating custom database product. Reverting to " + dbt, e);
+            }
+          }
+          else {
+            LOG.warn("Unexpected: metastore.use.custom.database.product was set, " +
+                     "but metastore.custom.database.product.classname was not. Reverting to " + dbt);
+          }
+        }
+
+        if (theDatabaseProduct == null) {
+          theDatabaseProduct = new DatabaseProduct();
```

Comment:
It looks like when this line is executed the dbType is null. Do you need to initialize a default value of dbType = UNDEFINED?

Reply:
The initialization happens at line 91. I don't see how we can get here without initialization.
LMK if I'm missing something.